File: chameleon_vpn_client/plugins/flutter_wireguard_plugin/macos/Classes/FlutterWireguardPlugin.swift
import Cocoa
import FlutterMacOS

public class FlutterWireguardPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private var channel: FlutterMethodChannel?
  private var eventChannel: FlutterEventChannel?
  private var statusSink: FlutterEventSink?
  private var configPath: String?
  private var isConnected = false

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_wireguard_plugin", binaryMessenger: registrar.messenger)
    let eventChannel = FlutterEventChannel(name: "flutter_wireguard_plugin/status", binaryMessenger: registrar.messenger)
    let instance = FlutterWireguardPlugin()
    instance.channel = channel
    instance.eventChannel = eventChannel
    registrar.addMethodCallDelegate(instance, channel: channel)
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("macOS " + ProcessInfo.processInfo.operatingSystemVersionString)
      
    case "connect":
      if let args = call.arguments as? [String: Any],
         let config = args["config"] as? String {
        let success = connectWireGuard(config: config)
        if success { statusSink?("connected") }
        result(success)
      } else {
        result(FlutterError(code: "BAD_ARGS", message: "Config param missing", details: nil))
      }
      
    case "disconnect":
      let success = disconnectWireGuard()
      if success { statusSink?("disconnected") }
      result(success)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
  private func connectWireGuard(config: String) -> Bool {
    let tempURL = FileManager.default.temporaryDirectory.appendingPathComponent("wg.conf")
    do {
      try config.write(to: tempURL, atomically: true, encoding: .utf8)
    } catch {
      NSLog("Failed to write config: \(error)")
      return false
    }
    let process = Process()
    process.launchPath = "/usr/local/bin/wg-quick"
    process.arguments = ["up", tempURL.path]
    process.launch()
    process.waitUntilExit()
    if process.terminationStatus == 0 {
      configPath = tempURL.path
      isConnected = true
      return true
    }
    return false
  }

  private func disconnectWireGuard() -> Bool {
    guard let path = configPath else { return false }
    let process = Process()
    process.launchPath = "/usr/local/bin/wg-quick"
    process.arguments = ["down", path]
    process.launch()
    process.waitUntilExit()
    configPath = nil
    isConnected = false
    return process.terminationStatus == 0
  }
  public func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
    statusSink = eventSink
    eventSink(isConnected ? "connected" : "disconnected")
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    statusSink = nil
    return nil
  }
}

File: chameleon_vpn_client/plugins/flutter_wireguard_plugin/android/src/main/kotlin/com/example/flutter_wireguard_plugin/FlutterWireguardPlugin.kt
package com.example.flutter_wireguard_plugin

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File

class FlutterWireguardPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
  private lateinit var channel : MethodChannel
  private lateinit var statusChannel: EventChannel
  private var events: EventChannel.EventSink? = null
  private lateinit var context: Context
  private var lastConfigFile: File? = null
  private var isConnected = false

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_wireguard_plugin")
    channel.setMethodCallHandler(this)
    statusChannel = EventChannel(flutterPluginBinding.binaryMessenger, "flutter_wireguard_plugin/status")
    statusChannel.setStreamHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "getPlatformVersion" -> {
        result.success("Android ${android.os.Build.VERSION.RELEASE}")
      }
      "connect" -> {
        val config = call.argument<String>("config")
        if (config == null) {
          result.error("INVALID_ARGUMENT", "Config param is null", null)
          return
        }
        val success = connectWireGuard(config)
        result.success(success)
      }
      "disconnect" -> {
        val success = disconnectWireGuard()
        result.success(success)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    this.events = events
    events?.success(if (isConnected) "connected" else "disconnected")
  }

  override fun onCancel(arguments: Any?) {
    events = null
  }

  private fun connectWireGuard(config: String): Boolean {
    return try {
      val file = File.createTempFile("wg", ".conf", context.cacheDir)
      file.writeText(config)
      val process = ProcessBuilder("sh", "-c", "wg-quick up ${file.absolutePath}").start()
      val exit = process.waitFor()
      if (exit == 0) {
        lastConfigFile = file
        isConnected = true
        events?.success("connected")
        true
      } else {
        file.delete()
        false
      }
    } catch (e: Exception) {
      e.printStackTrace()
      false
    }
  }

  private fun disconnectWireGuard(): Boolean {
    val file = lastConfigFile ?: return false
    return try {
      val process = ProcessBuilder("sh", "-c", "wg-quick down ${file.absolutePath}").start()
      val exit = process.waitFor()
      file.delete()
      lastConfigFile = null
      val success = exit == 0
      if (success) {
        isConnected = false
        events?.success("disconnected")
      }
      success
    } catch (e: Exception) {
      e.printStackTrace()
      false
    }
  }
}
